### PR TITLE
[FIX] purchase_stock: include qty in lead days computation

### PR DESCRIPTION
**Pull request by adwid**
The combination of a RR and a supplier that has a minimum quantity and a
lead time leads to incorrect results

To reproduce the issue:
1. Create a product P:
    - Type: Storable
2. Add a reordering rule for P (min: 10, max: 50)
3. Add a vendor to P
    - Quantity: 1
    - Price: 1
    - Delivery Lead Time: 7
4. Run the scheduler
5. Open the generated PO

Error: The order deadline is <today minus 7 days> instead of today. The
receipt date is today instead of <today plus 7 days>

When computing the quantity to order, the method needs the value of
`qty_forecast` (L254 in [1]) The computation of `qty_forecast` leads to
the computation of `lead_days_date`. To do so, it gets the lead days of
the associated stock rule `_get_lead_days`. In this method, a seller is
selected: [2] However, none of the parameter is defined: [3]

In the above case, the seller has a minimum quantity of 1. Therefore,
since this information is not given in [2], the vendor is not selected
and his lead time is therefore not taken into account. So,
`lead_days_date` will be equal to 0 and won't be recomputed. As a
result, the planned delivery date is today. Then, the quantity to order
is computed (L260 in [1]), so later, the module is able to select the
correct seller. However, this one has a lead time of 7 days -> the
generated PO will have the planned date to today and the order deadline
to <today minus 7 days>

Once an orderpoint has its quantity to order defined, its lead time
should be recomputed because it may be different depending to the
selected seller (and thus we should take this quantity into account when
searching for a seller)

[1]
https://github.com/odoo/odoo/blob/59ac8d891de6e20604494d6cfc6fc357a88d226c/addons/stock/models/stock_orderpoint.py#L254-L260
[2]
https://github.com/odoo/odoo/blob/1b03087524080f1a9751b1561cd8891466e72367/addons/purchase_stock/models/stock_rule.py#L153
[3]
https://github.com/odoo/odoo/blob/045c6ad3353bb0849d3f93d3488d8dcd8b8975b0/addons/product/models/product.py#L602

OPW-2614798

### CODE REVIEW

**Comment by adwid** [comment]
An alternative solution could be to edit the method signature of `_get_lead_days` so that, in `_compute_lead_days`, we can directly pass the quantity and the UoM to the method. On one hand it would be cleaner, but on the other it would be more invasive.

**Comment by amoyaux** [comment]
I would just use _prepare_sellers instead of _select_seller.
I agree it would be a little bit less accurate. However the qty_to_order already depends of lead_time so it would create a loop in the compute that I prefer avoid.

Since you need to know the delivery lead time to have an idea of the quantity missing at today + lead time -> if this quantity is not cover by a vendor what are we suppose to do? Check all the quantity by suppliers according their lead times (too costly) and hope having one that match

To support that at the better way I would instead do a feedack for master in order to think about it

**Comment by adwid** [comment]
I think we are falling into a technical limitation, don't you ?
The arbitrary selection of a seller will also lead to incorrect result, unfortunately
Suppose, before the step 3, I add a first seller (qty 1, price 10, lead time 1), the order deadline will be <today minus 6 days>

File: addons/purchase_stock/models/stock_rule.py [hunk 1]
```diff
@@ -147,10 +147,11 @@ def _get_lead_days(self, product):
         purchase and company lead time are always displayed for onboarding
         purpose in order to indicate that those options are available.
         """
+        ctx = self.env.context
         delay, delay_description = super()._get_lead_days(product)
-        bypass_delay_description = self.env.context.get('bypass_delay_description')
+        bypass_delay_description = ctx.get('bypass_delay_description')
         buy_rule = self.filtered(lambda r: r.action == 'buy')
-        seller = product.with_company(buy_rule.company_id)._select_seller()
+        seller = product.with_company(buy_rule.company_id)._select_seller(quantity=ctx.get('force_seller_qty', 0), uom_id=ctx.get('force_seller_uom', False))
```
[the remark this answers is not in the document]
Comment (adwid) [reply addons/purchase_stock/models/stock_rule.py:154]: I think that while we don't provide the correct qty (which seems tricky due to inter-dependencies), we will always find a use case that does not work
Considering your suggestion, suppose we have two vendors:
| Vendor | Quantity | Price | Lead time |
|--------|:--------:|:-----:|:---------:|
| 01     |    100   |   8   |     1     |
| 02     |     5    |   10  |     7     |

with a RR min:10-max:50, the same issue will appear again

[the remark this answers is not in the document]
Comment (ryv-odoo) [reply addons/purchase_stock/models/stock_rule.py:154]: Yep you are right, there isn't a perfect solution here.

With your change, you manage only simple case (one depend cycle between the qty to order and lead time) right (If I understand well) ? 

In master: we should compute lead_days at the same time than qty_to_order and filter-out impossible solution (qty_order, seller.lead_days). But it asks a big refactor of orderpoints (and _get_lead_days).

In Stable, we need to decide which solution is the best and simple enough to not break current flow of client.

**Comment by adwid** [comment]
@ryv-odoo I agree your suggestion should cover several cases
Considering the two tickets linked to this PR, the products only have one supplier so it will work
Let's hop it will cover most of our clients' configuration

Side note: using `float('inf')` doesn't work, at some point it leads to
```py
  File "/home/odoo/src/odoo/odoo/tools/float_utils.py", line 19, in round
    roundf = builtins.round(f)
OverflowError: cannot convert float infinity to integer
```
So I wanted to use the largest integer. However, after doing some research, it seems that such a value does not exist and it is suggested to use `maxsize` instead..

**Comment by ryv-odoo** [comment]
> @ryv-odoo I agree your suggestion should cover several cases
> Considering the two tickets linked to this PR, the products only have one supplier so it will work
> Let's hope it will cover most of our clients' configuration
> 
> Side note: using `float('inf')` doesn't work, at some point it leads to
> 
> ```python
>   File "/home/odoo/src/odoo/odoo/tools/float_utils.py", line 19, in round
>     roundf = builtins.round(f)
> OverflowError: cannot convert float infinity to integer
> ```
> 
> So I wanted to use the largest integer. However, after doing some research, it seems that such a value does not exist and it is suggested to use `maxsize` instead...

I would like avoid this because of the security flag (even if it is a false positive). Also, float can be bigger (well I shouldn't happen in a real case but we never know).
The maximum of the a double presicion (psql column type of min_qty) is `1.7976931348623157e+308` : 
```python
product.with_company(buy_rule.company_id)._select_seller(quantity=1.7976931348623157e+308)
```
It is quite hardcoded, but it should work.

**Comment by adwid** [comment]
Unfortunately, using `1.7976931348623157e+308` triggers the same error
I was looking for the largest integer value in PSQL and I found that the largest value of a `bigint` is `9223372036854775807` which is actually the value of `maxsize`. I know floats can be bigger, but honestly I still doubt that this situation will take place.

**Comment by ryv-odoo** [comment]
> Unfortunately, using `1.7976931348623157e+308` triggers the same error
> I was looking for the largest integer value in PSQL and I found that the largest value of a `bigint` is `9223372036854775807` which is actually the value of `maxsize`. I know floats can be bigger, but honestly I still doubt that this situation will take place.

Mhhhh, I am not fan of this. @Whenrow proposes a other solution less tricky (IMO): accept None as quantity in _select_seller and change the _select_seller to take in account this and bypass the check of quantity in this case. What do you think ?

**Comment by adwid** [comment]
I think then it would be better to use Arnold's suggestion
```diff
diff --git a/addons/purchase_stock/models/stock_rule.py b/addons/purchase_stock/models/stock_rule.py
index 33144e9e034..f30561953d7 100644
--- a/addons/purchase_stock/models/stock_rule.py
+++ b/addons/purchase_stock/models/stock_rule.py
@@ -150,7 +150,7 @@ class StockRule(models.Model):
         delay, delay_description = super()._get_lead_days(product)
         bypass_delay_description = self.env.context.get('bypass_delay_description')
         buy_rule = self.filtered(lambda r: r.action == 'buy')
-        seller = product.with_company(buy_rule.company_id)._select_seller()
+        seller = product.with_company(buy_rule.company_id)._prepare_sellers()[:1]
         if not buy_rule or not seller:
             return delay, delay_description
         buy_rule.ensure_one()
```
Otherwise we would have:
```diff
diff --git a/addons/product/models/product.py b/addons/product/models/product.py
index 33b02f1cba3..fa63a563d5c 100644
--- a/addons/product/models/product.py
+++ b/addons/product/models/product.py
@@ -610,7 +610,7 @@ class ProductProduct(models.Model):
         sellers = sellers.filtered(lambda s: not s.company_id or s.company_id.id == self.env.company.id)
         for seller in sellers:
             # Set quantity in UoM of seller
-            quantity_uom_seller = quantity
+            quantity_uom_seller = quantity or 0.0
             if quantity_uom_seller and uom_id and uom_id != seller.product_uom:
                 quantity_uom_seller = uom_id._compute_quantity(quantity_uom_seller, seller.product_uom)
 
@@ -620,7 +620,7 @@ class ProductProduct(models.Model):
                 continue
             if partner_id and seller.name not in [partner_id, partner_id.parent_id]:
                 continue
-            if float_compare(quantity_uom_seller, seller.min_qty, precision_digits=precision) == -1:
+            if float_compare(quantity_uom_seller, seller.min_qty, precision_digits=precision) == -1 and quantity is not None:
                 continue
             if seller.product_id and seller.product_id != self:
                 continue
diff --git a/addons/purchase_stock/models/stock_rule.py b/addons/purchase_stock/models/stock_rule.py
index 33144e9e034..6fa621ade6f 100644
--- a/addons/purchase_stock/models/stock_rule.py
+++ b/addons/purchase_stock/models/stock_rule.py
@@ -150,7 +150,7 @@ class StockRule(models.Model):
         delay, delay_description = super()._get_lead_days(product)
         bypass_delay_description = self.env.context.get('bypass_delay_description')
         buy_rule = self.filtered(lambda r: r.action == 'buy')
-        seller = product.with_company(buy_rule.company_id)._select_seller()
+        seller = product.with_company(buy_rule.company_id)._select_seller(quantity=None)
         if not buy_rule or not seller:
             return delay, delay_description
         buy_rule.ensure_one()
```
because with the first diff, we keep the clarity of the method `_select_seller` and its `quantity` parameter. Also, the result is quite the same in both cases: arbitrary selection of a seller (except that in the second diff, we use the cheapest seller, but not sure this selection is more relevant ?)

**Comment by ryv-odoo** [comment]
Yeah you are right.

about the "we use the cheapest seller, but not sure this selection is more relevant", it is only true if the cheaper seller partner0 is the same than the first one (due to the `res.name == seller.name` in the last if).

**Comment by adwid** [comment]
I think I spoke too soon, using the `_prepare_seller` solution would actually bring back another issue (c1bd985ef6a77c8848c4a2a8b92508511d9d8638)
The fix is updated

**Comment by amoyaux** [comment]
robodoo r+

**Pull request closed without merging**